### ISSUE TITLE
Array{Int,3}() should create an empty 3d array

### ISSUE THREAD

**Issue by eschnett** [JuliaLang/julia#14201]
`Array{Int,1}()` and `Array{Int,2}()` create empty 1d- and 2d-arrays, respectively. This could also be implemented for all higher dimensions.

**Comment by Rory-Finnegan** [comment]
+1 I ran into this last week and had to do `Array(Int, 2, 3, 4)` instead.

**Comment by yuyichao** [comment]
Note that high dimension arrays cannot be resized so it's unclear what is `Array{T,N}()` with `N >= 2` good for. On the other hand, I think it's useful to allow `Array{Int,3}(2, 3, 4)` in additional to `Array{Int}(2, 3, 4)` and `Array(Int, 2, 3, 4)`.

@Rory-Finnegan If you want `Array(Int, 2, 3, 4)` an empty 3-d array won't be useful for you.

**Comment by Rory-Finnegan** [comment]
@yuyichao Sorry, my initial use case was an empty 3-d array, but when I couldn't figure out how to do that I just preallocated a large one and used `sub` once I had filled it in (not the best solution, but hey). In either case, it did surprise me that `Array{Int,3}()` and `Array{Int,3}(2,3,4)` weren't valid constructors.

**Comment by JeffBezanson** [comment]
Honestly I'm not a fan of `Array{Int,2}()` making a 0x0 array. I find that both non-obvious and useless. +1 to the versions that accept size arguments though.

**Comment by GunnarFarneback** [comment]
> Note that high dimension arrays cannot be resized so it's unclear what is Array{T,N}() with N >= 2 good for.

Type-stable placeholders.

**Comment by nalimilan** [comment]
+1, at the very least to ensure consistency.

**Comment by stevengj** [comment]
Consistency with what?

**Comment by KristofferC** [comment]
With N=1 and N=2?

**Comment by StefanKarpinski** [comment]
I'm not sure why this works for N=1 and N=2 in the first place. That seems like a misfeature to me. It's as if we've decided that the default size of dimensions is zero but only if we don't specify any dimensions. What?!?! That strikes me as absurd and I'm somewhat tempted to find out when we started supporting this. (It's probably my fault, lol.) For N=0 it makes sense since there are no dimensions to specify.

**Comment by StefanKarpinski** [comment]
The constructors with specified dimensions already all works:

```jl
Array{Int,3}((2,3,4))
Array{Int}((2,3,4))
Array{Int,3}(2,3,4)
Array{Int}(2,3,4)
```

**Comment by StefanKarpinski** [comment]
The N=1 case actually kind of makes sense since we can at least grow those.

**Comment by martinholters** [comment]
The problem I have with the `N==1` case is that it makes one think that if `Vector{Int}()` is equivalent to `Int[]`, then `Vector{Int}(42)` is equivalent to `Int[42]`.

I agree that as a beginner, `Vector{Int}()` is one of the first things to try to create an empty vector. But you better learn using `[]` quickly anyway, as you'll need it for non-empty vectors of all kinds (concatenation, comprehension).

**Comment by StefanKarpinski** [comment]
I'm more ok with `Vector{Int}()` and `Vector()` because zero seems like a sensible default for the length of a growable vector, but I agree that there's a solid argument to be made for its removal.

**Comment by KristofferC** [comment]
Closed by #20330
